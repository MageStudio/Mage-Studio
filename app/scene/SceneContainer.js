import React from 'react';
import {
    connect
} from 'react-redux';

import {
    meshChanged,
    meshAttached,
    meshDetached
} from '../actions/currentMesh';

import {
    saveScene,
    loadScene,
    startProject,
    stopProject
} from '../actions/scene';

import Scene from './Scene';
import SceneToolbar from './SceneToolbar';
import Player from './Player';
import ModelModal from '../modals/ModelUploadModal';

import './scene.scss';
import { showModelUploadModal } from '../actions/models';

class SceneContainer extends React.Component {

    constructor(props) {
        super(props);

        this.state = {
            fullscreen: false
        };
    }

    handleFullScreen = (fullscreen) => () => {
        this.setState({ fullscreen });
    }

    render() {
        const {
            scene,
            store,
            config,
            loadScene,
            onMeshChanged,
            onMeshAttached,
            onMeshDetached,
            onSceneExported,
            startProject,
            showModelModal,
            stopProject
        } = this.props;

        const { fullscreen } = this.state;

        const { projectPlayerVisible, projectUrl } = scene;
        const fullScreenClassName = fullscreen ? 'fullscreen' : '';
        const className = `scene-container ${fullScreenClassName}`;

        return (
            <div className={className}>
                <Player
                    url={projectUrl}
                    visible={projectPlayerVisible} />
                <Scene
                    fullscreen={fullscreen}
                    config={config}
                    store={store}
                    onSceneLoad={loadScene}
                    onMeshChanged={onMeshChanged}
                    onMeshAttached={onMeshAttached}
                    onMeshDetached={onMeshDetached}
                    onSceneExported={onSceneExported} />
                <SceneToolbar
                    playerVisible={projectPlayerVisible}
                    stopProject={stopProject}
                    config={config}
                    showModelModal={showModelModal}
                    startProject={startProject}
                    fullscreen={fullscreen}
                    onFullScreen={this.handleFullScreen}/>
                <ModelModal />
            </div>
        );
    }
}

const mapStateToProps = ({ config, scene }) => ({
    config,
    scene
});

const mapDispatchToProps = (dispatch) => ({

    startProject: project => dispatch(startProject(project)),
    showModelModal: () => dispatch(showModelUploadModal()),
    onMeshChanged: ({element, position, rotation, scale}) => {
        return dispatch(meshChanged(element, position, rotation, scale))
    },
    onMeshAttached: ({element, position, rotation, scale}) => {
        return dispatch(meshAttached(element, position, rotation, scale))
    },
    onMeshDetached: () => dispatch(meshDetached()),
    onSceneExported: (name) => ({ data }) => dispatch(saveScene(name, data)),
    loadScene: (name) => dispatch(loadScene(name)),
    stopProject: (project) => dispatch(stopProject(project))
});

export default connect(mapStateToProps, mapDispatchToProps)(SceneContainer);
