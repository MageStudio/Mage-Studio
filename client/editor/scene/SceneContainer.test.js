import React from 'react';
import { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';
import sinon from 'sinon';
import { SceneContainer } from './SceneContainer';

const getSceneContainer = (props) => shallow(<SceneContainer scene={{}} {...props} />);

describe('SceneContainer', () => {

    it('should render properly', () => {
        const component = getSceneContainer();

        expect(toJSON(component)).toMatchSnapshot();
    });

    it('should render properly if fullscreen', () => {
       const component = getSceneContainer();

       component.setState({ fullscreen: true });

       expect(toJSON(component)).toMatchSnapshot();
    });

    it('should have state.fullscreen set to false', () => {
        const component = getSceneContainer();

        expect(component.state().fullscreen).toBe(false);
    });

    describe('events', () => {

        it('should call onMeshChanged', () => {
            const spy = sinon.spy();
            const component = getSceneContainer({ onMeshChanged: spy });

            component.find('Scene').simulate('meshChanged');

            expect(spy.called).toBe(true);
        });

        it('should call onMeshAttached', () => {
            const spy = sinon.spy();
            const component = getSceneContainer({ onMeshAttached: spy });

            component.find('Scene').simulate('meshAttached');

            expect(spy.called).toBe(true);
        });

        it('should call onMeshDetached', () => {
            const spy = sinon.spy();
            const component = getSceneContainer({ onMeshDetached: spy });

            component.find('Scene').simulate('meshDetached');

            expect(spy.called).toBe(true);
        });

        it('should call onSceneExported', () => {
            const spy = sinon.spy();
            const component = getSceneContainer({ onSceneExported: spy });

            component.find('Scene').simulate('sceneExported');

            expect(spy.called).toBe(true);
        });
    });
});