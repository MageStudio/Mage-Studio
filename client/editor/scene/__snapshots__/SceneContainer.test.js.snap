// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SceneContainer should render properly 1`] = `
<div
  className="scene-container"
>
  <Player />
  <Scene
    fullscreen={false}
  />
</div>
`;

exports[`SceneContainer should render properly if fullscreen 1`] = `
<div
  className="scene-container fullscreen"
>
  <Player />
  <Scene
    fullscreen={true}
  />
</div>
`;
